### DISCUSSION TITLE
Got Remote Desktop working via KMS capture

### DISCUSSION THREAD

**Discussion by thepragmaticmero** [YaLTeR/niri#680] [Show and tell]
Well, there's no way to capture via VNC, RDP, RustDesk, etc, at the moment. Until daddy YaLTeR or some brave rust wayland dev implements it. So for now the only way I managed to do remote desktop is by bypassing wayland completely :D 

#### Server (with niri wayland archbtw)
- [Sunshine](https://github.com/LizardByte/Sunshine)
#### Client (with whatever)
- [Moonlight](https://flathub.org/apps/com.moonlight_stream.Moonlight)

This is a guide for Arch BTW cus ArchBTW. 
- Follow [this guide](https://github.com/LizardByte/pacman-repo) to add the sunshine repo. Or install using a AUR helper `paru -Syu sunshine-bin`
- Install sunshine `sudo pacman -Syu sunshine` 
- Enable avahi-daemon first, otherwise you'll get an error `systemctl enable --now avahi-daemon`
- Give sunshine root permission to KMS capture (bypassing wayland) `sudo setcap cap_sys_admin+p $(readlink -f $(which sunshine))`
- Enable the sunshine service as user `systemctl --user enable --now sunshine`
- Go to https://localhost:[port] and setup username and password for sunshine.
- Go to your client and put your server's ip on moonlight, install with [this guide](https://github.com/moonlight-stream/moonlight-qt?tab=readme-ov-file#downloads), check the ip on the server with `ip a`, moonlight might detect it automatically.
- Pair the pin on the server.
- That's it I guess... I'm too lazy to finish writing this correctly. I'm going to sleep. zzzzzzz.

> Lesson of the day: Writing documentation sucks, but its so important tho. So do it.

Demo image:
![image](https://github.com/user-attachments/assets/092d0875-dc7c-4128-b7f5-4bb41d92b956)

**Comment by luochen1990** [comment]
As I know, Sunshine needs a phisical monitor, what if I don't have a monitor connected?

**Reply by andar1an** [reply]
You can get dummy adapters if you have HDMI or Display Port.

**Reply by iraizo** [reply]
There's also Windows drivers that emulate a monitor for this specific use case, I'm sure you will find something similar done for Linux

**Comment by Yilmaz4** [comment]
I keep getting "RTSP handshake failed with error 60", and I see `Error: Initial Ping Timeout`" in sunshine logs. This only happens on Niri, it works on Hyprland. I've confirmed that sunshine is using KMS.

**Reply by Yilmaz4** [reply]
Update: running sunshine as root fixes this, but then sound doesn't work

**Reply by Yilmaz4** [reply]
Update: I fixed that by setting the environment variables `XDG_RUNTIME_DIR=/run/user/1000` and `PULSE_SERVER=unix:/run/user/1000/pulse/native` for the sh running sunshine

**Reply by kamiicode** [reply]
you are my hero

**Comment by SmallStrayFox** [comment]
has anyone managed to get keyboard shortcuts in niri working via vnc? i can connect via vnc to a host running wayvnc and niri, but keyboard shortcuts get eaten up. i can see all the keys are showing correctly when i hit them in wev to test they are being passed through by the vnc client, but when i hit e.g. super+up arrow it will print the letter A in the terminal instead. notably the up arrow ansi escape sequence is ESC [ A. so somewhere between the vnc client and niri seeing the keyboard shortcut is getting altered. i can get moonlight working, but if it's possible to use vnc too that would be useful, and it seems to just be shortcuts not working, but that's kinda important with niri.

**Reply by akhiljalagam** [reply]
> has anyone managed to get keyboard shortcuts in niri working via vnc? i can connect via vnc to a host running wayvnc and niri, but keyboard shortcuts get eaten up. i can see all the keys are showing correctly when i hit them in wev to test they are being passed through by the vnc client, but when i hit e.g. super+up arrow it will print the letter A in the terminal instead. notably the up arrow ansi escape sequence is ESC [ A. so somewhere between the vnc client and niri seeing the keyboard shortcut is getting altered. i can get moonlight working, but if it's possible to use vnc too that would be useful, and it seems to just be shortcuts not working, but that's kinda important with niri.

I use sunshine and moonlight for remote access and its working without any issues

**Reply by SmallStrayFox** [reply]
i've successfully got moonlight working, but as far as i can see right now sunshine on linux does not support yuv444 and so text on screen just feels slightly off, desktops dont feel quite as crisp as i'd like. moonlight is 99% of the way there, but 4:4:4 chroma subsampling would be the final 1%. as it is, i was trying to see if wayvnc would work as an alternative remote desktop, and it definitely does, but keyboard shortcuts seem to get eaten up somewhere between vnc and the compositor niri. i will continue to use moonlight if nothing else works, i just dont like how it captures the mouse as compared to a vnc session.

**Reply by akhiljalagam** [reply]
Yeah. I tried WayVNC with RealVNC client. I faced mouse issues and the Alt key didn't work. But with Moonlight, I didn't face any blur or latency issues. It's actually working better. I am on a 5G LAN, by the way.

**Reply by Atan-D-RP4** [reply]
All compositor keybinds will fail. It currently does not work because of an incomplete implementation of virtual keyboard recently broke, and was thus reverted, and is currently waiting on upstream smithay work for a complete implementation.

**Comment by Norgus** [comment]
Thanks for the no-nonsense guide!
In case anyone is using wireguard, or another home VPN solution, and like me wondering why everything else works through their tunnel except Sunshine <-> Moonlight -
I ensured my firewall was allowing all the TCP and UDP ports necessary & then just moved WG to my firewall's trusted zone (everything open) - still not working!
I had forgotten that I used an option in wireguard on Android to only apply the connection to specific apps! As soon as I added Moonlight to said list, of course it worked.

**Comment by thesomewhatyou** [comment]
Hahaha... my mouse doesn't work. Nor does my keyboard.
